postgres_replicator/transfer.go

```diff
@@ -0,0 +1,228 @@
+// transfers.
+func (t *Transfer) RunContinuously(sleepAfterTransferInSecs int) {
+	for {
+		t.Run()
+		log.Printf("Sleeping for %d seconds", sleepAfterTransferInSecs)
+		time.Sleep(time.Duration(sleepAfterTransferInSecs) * time.Second)
+	}
+}
+
+func (t *Transfer) transferTable(bigQueryDataset, tableName, dateField string, done chan bool) {
+	logger := NewLogger(tableName)
+
+	// Get the BigQuery table schema
+	bqSchema, err := t.bq.GetTableSchema(bigQueryDataset, tableName)
+	if err != nil {
+		logger.Errorf("Could not get BigQuery table schema: %v", err)
+		done <- true
+		return
+	}
+
+	// Convert BigQuery schema to Postgres schema
+	pgSchema, err := t.convertSchema(bqSchema)
+	if err != nil {
+		logger.Errorf("Could not convert schema: %v", err)
+		done <- true
+		return
+	}
+
+	// Create PostgreSQL table if needed
+	err = t.prepareTable(tableName, pgSchema)
+	if err != nil {
+		logger.Errorf("Could not prepare Postgres table: %v", err)
+		done <- true
+		return
+	}
+
+	// Get rows to transfer
+	rows, err := t.getBigQueryRows(bigQueryDataset, tableName, dateField, bqSchema)
+	if err != nil {
+		logger.Errorf("Could not get data from BigQuery: %v", err)
+		done <- true
+		return
+	}
+
+	// Transfer rows to Postgres
+	err = t.transferToPostgres(tableName, pgSchema, rows, logger)
+	if err != nil {
+		logger.Errorf("Could not transfer one or more rows to Postgres: %v. ", err)
+		done <- true
+		return
+	}
+
+	done <- true
+}
+
+func (t *Transfer) convertSchema(bqSchema *BigQuerySchema) (*PostgresSchema, error) {
+	pgSchema := &PostgresSchema{make(map[string]string)}
+	for columnName, dataType := range bqSchema.schema {
+		if strings.Contains(dataType, "STRUCT") {
+			pgSchema.schema[columnName] = "JSON"
+			continue
+		}
+		if strings.Contains(dataType, "FLOAT64") {
+			pgSchema.schema[columnName] = "DOUBLE PRECISION"
+			continue
+		}
+		if strings.Contains(dataType, "STRING") {
+			pgSchema.schema[columnName] = "TEXT"
+			continue
+		}
+		if strings.Contains(dataType, "TIME") {
+			pgSchema.schema[columnName] = "TIMESTAMPTZ"
+			continue
+		}
+		pgSchema.schema[columnName] = dataType
+	}
+	return pgSchema, nil
+}
+
+func (t *Transfer) prepareTable(tableName string, pgSchema *PostgresSchema) error {
+	tableExists := t.pg.TableExists(tableName)
+	if tableExists {
+		return nil
+	}
+
+	err := t.pg.CreateTableFromSchema(tableName, pgSchema)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *Transfer) getBigQueryRows(bigQueryDataset, tableName, dateField string, bqSchema *BigQuerySchema) (*bigquery.RowIterator, error) {
+	// Get most recent entry from Postgres table
+	timestamp, err := t.pg.GetMostRecentEntry(tableName, dateField)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get most recent Postgres timestamp: %s", err)
+	}
+
+	// Get data after this time, or all data if last timestamp doesn't exist
+	rows, err := t.bq.GetDataAfterDatetime(bigQueryDataset, tableName, dateField, timestamp, bqSchema)
```

Comment:
if you're really planning to load all the newly added data in memory (which can be a lot btw) you should at least add a comment/todo why this is ok (you're assuming that this is only used for benchmark results and that there is only going to be a small number of them since you sync frequently).
In general this is not ok (you should at least do updates in batches).

Reply:
I was worried about this too. It seems like the Golang BigQuery library will [automatically page for you](https://cloud.google.com/bigquery/docs/paging-results)
> The Cloud Client Libraries for Go automatically paginates by default, so you do not need to implement pagination yourself

Just to be sure, I used [this method](https://stackoverflow.com/a/774601) to measure peak memory usage of the replicator when running in App Engine. When transferring a ~9GB table, peak memory usage over several samples was reported to be about 100MB.

I'll add a comment to the function